### PR TITLE
Implement model to calculate edge impurity concentration

### PR DESCRIPTION
**Pull request by tbody-cfs**
This merge request adds in the "Lengyel model", as described in
```
Nucl. Fusion 61 (2021) 046029 (21pp) https://doi.org/10.1088/1741-4326/abe4b2
"Comparison between SOLPS-4.3 and the Lengyel Model for ITER baseline neon-seeded plasmas"
D. Moulton, P.C. Stangeby, X. Bonnin and R.A. Pitts
```
The Lengyel model was originally presented in a 1981 IPP report "Analysis of Radiating Plasma Boundary Layers" by L. L. Lengyel, report number 1/191

The model gives the concentration of a radiating impurity required to achieve a requested radiative power dissipation fraction in the SOL. Usually, the calculated concentration can be related to the concentration from a higher-fidelity model such as SOLPS (see Moulton paper) via a constant factor, which we call the `lengyel_overestimation_factor`.

### CODE REVIEW

**Comment by tbody-cfs** [comment]
For the variables previously killing the pipeline, this shows the absolute difference (ref - test) and relative difference (1 - ref/test). rtol of 1e-5 (default for `xr.testing.assert_allclose` is large enough that the test passes, while `1E-6` doesn't. Still don't have a good idea why these differences are here, but they might be related to machine architecture
image
image
image

File: .github/workflows/workflow_actions.yml [hunk 1]
```diff
@@ -26,7 +26,7 @@ jobs:
       - uses: actions/checkout@v3
 
       - name: Install Poetry
-        run: curl -sSL https://install.python-poetry.org | python - --version 1.6.1
```
Comment (hassec) [inline .github/workflows/workflow_actions.yml:29]: Why was this removed? We want to ensure that the CI uses a consistent version.

File: cfspopcon/algorithms/__init__.py [hunk 2]
```diff
@@ -53,22 +55,6 @@ def get_algorithm(algorithm: Union[Algorithms, str]) -> Union[Algorithm, Composi
...
-    "calc_ohmic_power",
-    "calc_peaked_profiles",
-    "calc_plasma_current_from_q_star",
-    "calc_power_balance_from_tau_e",
-    "predictive_popcon",
-    "calc_q_star_from_plasma_current",
-    "two_point_model_fixed_fpow",
-    "two_point_model_fixed_qpart",
-    "two_point_model_fixed_tet",
-    "calc_zeff_and_dilution_from_impurities",
```
Comment (hassec) [inline cfspopcon/algorithms/__init__.py:71]: This means non of these show up by default in the docs. 
Why remove them here?

File: cfspopcon/formulas/scrape_off_layer_model/edge_impurity_concentration.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    L_int_integrator: Callable[[Unitfull, Unitfull], Unitfull],
+    q_parallel: Unitfull,
+    SOL_power_loss_fraction: Unitfull,
+    target_electron_temp: Unitfull,
+    upstream_electron_temp: Unitfull,
+    upstream_electron_density: Unitfull,
+    kappa_e0: Unitfull,
+    lengyel_overestimation_factor: Unitfull,
+) -> Unitfull:
+    """Calculate the relative concentration of an edge impurity required to achieve a given SOL power loss fraction.
```
Comment (hassec) [inline cfspopcon/formulas/scrape_off_layer_model/edge_impurity_concentration.py:62]: Could we add a reference here?

File: example_cases/SPARC_PRD/input.yaml [hunk 4]
```diff
@@ -166,3 +168,10 @@ fraction_of_P_SOL_to_divertor: 0.6
 kappa_e0: 2600.0
 # Calculate P_rad_SOL such that the target electron temperature in eV reaches this value (for FixedTargetElectronTemp)
 target_electron_temp: 25.0
+
+# Which species should be injected to cool the scrape-off-layer?
+edge_impurity_species: Argon
+# Factor applied to the result of the Lengyel model to give an absolute low-Z impurity concentration.
+lengyel_overestimation_factor: 4.3
+# Ratio of edge impurity concentration in the edge to in the core (c_edge / c_core)
+edge_impurity_enrichment: 5.0
```
Comment (hassec) [inline example_cases/SPARC_PRD/input.yaml:177]: Should `SPARC_PRD` evolve and have these added features or be fixed to use the same methods it originally was defined with?

File: pyproject.toml [hunk 5]
```diff
@@ -26,36 +26,35 @@ popcon = 'cfspopcon.cli:run_popcon_cli'
...
+types-pyyaml = ">=6.0.12.2"
+pandas-stubs = ">=1.5.1.221024"
+mypy = ">=1.4.1"
+sphinx = ">=7.2.6"
+sphinx-rtd-theme = ">=2.0.0rc2"
+sphinxcontrib-bibtex = ">=2.6.1"
+sphinx-copybutton = ">=0.5.2"
+ruff = ">=0.0.292"
+nbmake = ">=1.4.3"
+nbsphinx = ">=0.9.3"
```
Comment (hassec) [inline pyproject.toml:57]: What motivated all these changes? In general I prefer to have feature changes separate from dependency changes/updates.

**Review by hassec — CHANGES_REQUESTED** [review]

File: example_cases/SPARC_PRD/input.yaml (hunk 4 above)
Comment (hassec) [reply example_cases/SPARC_PRD/input.yaml:177]: Ok to move this to a new issue 👍 Would you mind creating that and linking it here?

File: cfspopcon/algorithms/__init__.py (hunk 2 above)
Comment (tbody-cfs) [reply cfspopcon/algorithms/__init__.py:71]: How do I do this?

File: example_cases/SPARC_PRD/input.yaml (hunk 4 above)
Comment (tbody-cfs) [reply example_cases/SPARC_PRD/input.yaml:177]: #27

**Review by hassec — APPROVED** [review]

**Comment by nelsonand** [comment]
@tbody-cfs What is the extent of this update? Does it help already to solve https://github.com/cfs-energy-internal/POPSIM/issues/28?

**Pull request merged**